kedro/io/kedro_data_catalog.py

```diff
@@ -0,0 +1,326 @@
+Then it will act as a single point of reference for your calls, relaying load and
+save functions to the underlying datasets.
+"""
+
+from __future__ import annotations
+
+import copy
+import difflib
+import logging
+import re
+from typing import Any
+
+from kedro.io.catalog_config_resolver import CatalogConfigResolver, Patterns
+from kedro.io.core import (
+    AbstractDataset,
+    AbstractVersionedDataset,
+    DatasetAlreadyExistsError,
+    DatasetError,
+    DatasetNotFoundError,
+    Version,
+    generate_timestamp,
+)
+from kedro.io.memory_dataset import MemoryDataset
+from kedro.utils import _format_rich, _has_rich_handler
+
+
+class KedroDataCatalog:
+    def __init__(
+        self,
+        datasets: dict[str, AbstractDataset] | None = None,
+        raw_data: dict[str, Any] | None = None,
+        config_resolver: CatalogConfigResolver | None = None,
+        load_versions: dict[str, str] | None = None,
+        save_version: str | None = None,
+    ) -> None:
+        """``KedroDataCatalog`` stores instances of ``AbstractDataset``
+        implementations to provide ``load`` and ``save`` capabilities from
+        anywhere in the program. To use a ``KedroDataCatalog``, you need to
+        instantiate it with a dictionary of data sets. Then it will act as a
+        single point of reference for your calls, relaying load and save
+        functions to the underlying data sets.
+
+        Args:
+            datasets: A dictionary of data set names and data set instances.
+            raw_data: A dictionary with data to be added in memory as `MemoryDataset`` instances.
+                Keys represent dataset names and the values are raw data.
+            config_resolver: An instance of CatalogConfigResolver to resolve dataset patterns and configurations.
+            load_versions: A mapping between data set names and versions
+                to load. Has no effect on data sets without enabled versioning.
+            save_version: Version string to be used for ``save`` operations
+                by all data sets with enabled versioning. It must: a) be a
+                case-insensitive string that conforms with operating system
+                filename limitations, b) always return the latest version when
+                sorted in lexicographical order..
+        """
+        self._config_resolver = config_resolver or CatalogConfigResolver()
+        self._datasets = datasets or {}
+        self._load_versions = load_versions or {}
+        self._save_version = save_version
+
+        self._use_rich_markup = _has_rich_handler()
+
+        for ds_name, ds_config in self._config_resolver.config.items():
+            self._add_from_config(ds_name, ds_config)
+
+        if raw_data:
+            self.add_raw_data(raw_data)
+
+    @property
+    def datasets(self) -> dict[str, Any]:
+        return copy.copy(self._datasets)
+
+    @datasets.setter
+    def datasets(self, value: Any) -> None:
```

Comment:
I think that's a good use case for [NoReturn](https://peps.python.org/pep-0484/#the-noreturn-type)?

```suggestion
    def datasets(self, value: Any) -> NoReturn:
```